File: leetcodes/143. Reorder List/reorderList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
  public ListNode midNode(ListNode head) {
    ListNode fast = head, slow = head;
    while(fast.next != null && fast.next.next != null) {
      fast = fast.next.next;
      slow = slow.next;
    }
    return slow;
  }
  public ListNode reverse(ListNode head){
    ListNode cur = head, prev = null, next = null;
    while(cur != null){
      next = cur.next;
      cur.next = prev;
      prev = cur;
      cur = next;
    }
    return prev;
  }
  public void reorderList(ListNode head) {
    //split the list by finding mid point
    //reverse another list
    //merge both list
    //time O(n) space O(1)
    ListNode midNode = midNode(head);
    ListNode nextToMid = midNode.next;
    midNode.next = null;
    ListNode p2 = reverse(nextToMid);
    ListNode p1 = head;
    ListNode p1Next;
    while(p1!= null && p2!= null) {
      p1Next = p1.next;
      p1.next = p2;
      p1 = p2;
      p2 = p1Next;
    }
  }
}

File: leetcodes/143. Reorder List/reorderList.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    m = 1
    '''
    timeO(n)-> reverse and get mid length
    SpaceO(n)->create new reverse list
    merge reverse and original list time(mid)
    '''
    def reverseList(self, head):
        cur = head
        prev = ListNode(cur.val, None)
        while cur.next:
          cur = cur.next
          copy = ListNode(cur.val, prev)
          prev = copy
          self.m += 1
        return prev
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        revHead = self.reverseList(head)
        cur = head
        next = None
        mid =self.m // 2
        while cur.next and revHead.next and mid:
          next = cur.next
          if (mid - 1) is 0 and revHead.val == next.val:
            cur.next = ListNode(revHead.val,None)
            break
          cur.next = ListNode(revHead.val, next)
          cur = next
          revHead = revHead.next
          if  (mid - 1) is 0 and revHead.val == next.val:
            next.next = None
            break
          mid -= 1
        return

#split to two list   and reverse the 2nd list , merge two list
# don't understand why it can find mid by two pointers.. maybe we can just count
class Solution:
    def reorderList(self, head):
      cur = head
      count = 1 #total is number counts
      while cur.next:
        cur = cur.next
        count +=1
      mid = count //2 #doesn't care odd or even
      sPoint = head
      while mid:
        sPoint = sPoint.next
        mid -= 1

      #reverse spoint list
      prev, curr = None, sPoint
      while curr:
        prev, curr.next, curr = curr, prev, curr.next
      #prev
      #Merge two lists in place
      first, second = head, prev
      while second.next:
        first.next, first = second, first.next
        second.next, second = first, second.next
      return
      #time O(n) space O(1)